File: apps/api/app/schemas/job_posting.py
from pydantic import BaseModel, HttpUrl
from typing import Optional, List, Dict, Any
from datetime import datetime
from app.schemas.application import ApplicationField


class JobIngestRequest(BaseModel):
    url: str


class RawExtraction(BaseModel):
    provider_payload: Optional[Dict[str, Any]] = None
    extraction_path: str
    fetched_at: str  # ISO datetime
    warnings: List[str] = []


class JobPostingResponse(BaseModel):
    id: str
    source_url: str
    provider: str
    company_name: Optional[str] = None
    title: Optional[str] = None
    location: Optional[str] = None
    employment_type: Optional[str] = None
    apply_url: Optional[str] = None
    description_html: Optional[str] = None
    description_text: str
    application_form_schema: Optional[List[ApplicationField]] = None
    raw: RawExtraction
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class JobIngestResponse(BaseModel):
    job_posting: JobPostingResponse
    status: str  # "success" | "needs_render" | "needs_extension_extraction"

File: packages/shared/src/schemas/jobPosting.ts
import { z } from 'zod';
import { ApplicationFieldSchema } from './application';

export const ProviderSchema = z.enum([
  'GREENHOUSE',
  'LEVER',
  'ASHBY',
  'SMARTRECRUITERS',
  'WORKDAY',
  'ORACLE_CX',
  'AVATURE',
  'SUCCESSFACTORS',
  'TALEO',
  'ICIMS',
  'PHENOM',
  'UNKNOWN',
]);

export const EmploymentTypeSchema = z.enum([
  'FULL_TIME',
  'PART_TIME',
  'CONTRACT',
  'TEMPORARY',
  'INTERNSHIP',
  'FREELANCE',
  'OTHER',
]);

export const RawExtractionSchema = z.object({
  provider_payload: z.record(z.any()).optional(),
  extraction_path: z.string(),
  fetched_at: z.string(), // ISO datetime
  warnings: z.array(z.string()),
});

export const JobPostingNormalizedSchema = z.object({
  source_url: z.string().url(),
  provider: ProviderSchema,
  company_name: z.string().optional(),
  title: z.string().optional(),
  location: z.string().optional(),
  employment_type: EmploymentTypeSchema.optional(),
  apply_url: z.string().url().optional(),
  description_html: z.string().optional(),
  description_text: z.string(), // Required if extraction succeeded
  application_form_schema: z.array(ApplicationFieldSchema).optional(),
  raw: RawExtractionSchema,
});

export type Provider = z.infer<typeof ProviderSchema>;
export type EmploymentType = z.infer<typeof EmploymentTypeSchema>;
export type RawExtraction = z.infer<typeof RawExtractionSchema>;
export type JobPostingNormalized = z.infer<typeof JobPostingNormalizedSchema>;
